### PR TITLE
[bugfix] Fix reading/writing of mesh layer to a qgis project #18801

### PR DESCRIPTION
**Pull request by PeterPetrik**
Added a new function `QgsMapLayer::encodeSource()` to clean the write XML function a bit. 

Fixes #18801

### CODE REVIEW

**Comment by PeterPetrik** [comment]
Fixed build errors, ready for review

File: src/core/qgsmaplayer.h [hunk 1]
```diff
@@ -1197,6 +1197,15 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
     virtual bool writeXml( QDomNode &layer_node, QDomDocument &document, const QgsReadWriteContext &context ) const;
 
+    /**
+     * Called by writeLayerXML(), used by children to encode provider's specific data
+     * source to project files
+     *
+     * \since QGIS 3.2
+     */
+    virtual QString encodeSource( const QgsReadWriteContext &context ) const;
```
Comment (wonder-sk) [inline src/core/qgsmaplayer.h]: I would suggest renaming to encodedSource(...)

File: src/core/qgsmaplayer.h [hunk 2]
```diff
@@ -1197,6 +1197,15 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
     virtual bool writeXml( QDomNode &layer_node, QDomDocument &document, const QgsReadWriteContext &context ) const;
 
+    /**
+     * Called by writeLayerXML(), used by children to encode provider's specific data
```
Comment (wonder-sk) [inline src/core/qgsmaplayer.h]: s/children/derived classes/

File: src/core/qgsproject.cpp [hunk 3]
```diff
@@ -820,6 +821,10 @@ bool QgsProject::addLayer( const QDomElement &layerElem, QList<QDomNode> &broken
   {
     mapLayer = new QgsRasterLayer;
   }
+  else if ( type == QLatin1String( "mesh" ) )
+  {
+    mapLayer = new QgsMeshLayer;
+  }
```
Comment (wonder-sk) [inline src/core/qgsproject.cpp]: Can we also have equivalent of encodedSource() for reading - e.g. decodedSource()

Comment (PeterPetrik) [reply src/core/qgsproject.cpp]: done

**Review by wonder-sk — APPROVED** [review]

File: src/core/mesh/qgsmeshlayer.cpp [hunk 4]
```diff
@@ -180,9 +181,84 @@ bool QgsMeshLayer::writeSymbology( QDomNode &node, QDomDocument &doc, QString &e
...
+
+  return mValid; // should be true if read successfully
+}
+
+bool QgsMeshLayer::writeXml( QDomNode &layer_node, QDomDocument &document, const QgsReadWriteContext &context ) const
+{
+  // first get the layer element so that we can append the type attribute
+  QDomElement mapLayerNode = layer_node.toElement();
+
+  if ( mapLayerNode.isNull() || ( "maplayer" != mapLayerNode.nodeName() ) )
```
Comment (m-kuhn) [inline src/core/mesh/qgsmeshlayer.cpp]: QLatin1String?

**Pull request merged**